src/machine/streams.rs

```diff
@@ -636,6 +636,20 @@ impl Stream {
         }
     }
 
+    pub(crate) fn is_closed(&self) -> bool {
+        match self.stream_inst.0.borrow_mut().stream_inst {
+            StreamInstance::Null => true,
+            StreamInstance::TcpStream(_, ref mut tcp_stream) => {
+                let mut buf = [0;8];
+                match tcp_stream.peek(&mut buf) {
+                    Ok(n_bytes) => n_bytes == 0,
```

Comment:
This seems somewhat suspect: If no bytes are pending at the moment, the test will succeed, also if the stream is not yet closed. Are you sure this is correct? It seems that the stream is not closed if we get `Ok(_)` here.

Reply:
Well, it's a bit complicated... (disclaimer: I'm not a networking expert, this is just from what I've researched)

There's no way to check if a TcpStream is closed other than interact with it. So, the easiest way to do it is just read it (actually not reading, as peek does not remove the bytes from the queue). If the stream is closed, it will return a Ok(0) immediately. But if it is not closed, it will block until a byte can be sent to the application (doesn't need to fill the whole buffer) [https://users.rust-lang.org/t/how-to-detect-tcp-close/50925]. Err is for other kind of errors in this context.

Now, what happens if we call shutdown in Rust and the stream is closed but there's no data left to write/read: it just works, it doesn't give an error because for all purposes, nothing more needs to be done. But we want to throw an exception when the stream is closed! The map between the real TCP streams and Prolog streams is not 1:1 and this edge case shows.

What I'm thinking is that if we're calling is_closed only in `close/1` (which will close the stream or throw an exception), maybe we can just try to write instead of reading and that would remove the possible block but I don't know if that will interfere in some protocols.

Or another option is to just call shutdown in Rust and throw an exception if Rust code returns Err, which means that will only fail if there was some data left, but it will not block nor interfere with protocols.

What do you think?

Reply:
Thank you for this explanation! You can prevent infinite blocking by setting a timeout on the stream:

https://stackoverflow.com/questions/33557375/how-do-i-prevent-tcpstream-from-blocking-on-a-read

Would this work?

I would definitely recommend *against* writing anything to the stream. In addition, I think a small client/server test case (written in Prolog) that creates a few connections, maybe writes a few things and receives them, and also closes the sockets would be great, so that the current problem can be reliably reproduced, and then shown to be corrected with this patch.

Reply:
It works, but it will say a stream is closed if it is not closed but there's no more data in the timeout.

Thinking and thinking again I think a is_closed function is a dead end, it will never work transparently to the protocol the app is using. I think the right choice is just to manage the shutdown errors (which does not fail if there's no data to read/write but it's better than nothing). By the way, using shutdown seems to be just a good practice, Rust will just close the stream as soon as it goes out of the scope.

I will provide some test cases

Reply:
I've updated the code with the new proposal. I'll submit the test code tomorrow as right now it's just a bunch of telnet's

Reply:
Awesome, thank you a lot!

Reply:
Submitted. Check the two different outputs:
```
aarroyoc@adrianistan ~/d/s/tests-pl (close-stream)> scryer-prolog issue1046-close-stream.pl 
?- test.
Listening at port 5000
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 107, kind: NotConnected, message: "Transport endpoint is not connected" }', src/machine/streams.rs:167:53
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
aarroyoc@adrianistan ~/d/s/tests-pl (close-stream) [101]> cargo run -- issue1046-close-stream.pl 
   Compiling scryer-prolog v0.8.128 (/home/aarroyoc/dev/scryer-prolog)
    Finished dev [unoptimized + debuginfo] target(s) in 8.17s
     Running `/home/aarroyoc/dev/scryer-prolog/target/debug/scryer-prolog issue1046-close-stream.pl`
?- test,
true.
Listening at port 5000
caught: error(existence_error(stream,'$stream'(0x55d2138df3c0)),close/1)
?- 
```